dbms/src/Interpreters/SystemLog.h

```diff
@@ -183,51 +176,86 @@ SystemLog<LogElement>::SystemLog(Context & context_,
 {
     log = &Logger::get("SystemLog (" + database_name + "." + table_name + ")");
 
-    data.reserve(DBMS_SYSTEM_LOG_QUEUE_SIZE);
-    saving_thread = ThreadFromGlobalPool([this] { threadFunction(); });
+    saving_thread = ThreadFromGlobalPool([this] { savingThreadFunction(); });
 }
 
 
 template <typename LogElement>
 void SystemLog<LogElement>::add(const LogElement & element)
 {
+    std::unique_lock lock(mutex);
+
     if (is_shutdown)
         return;
 
-    /// Without try we could block here in case of queue overflow.
-    if (!queue.tryPush({EntryType::LOG_ELEMENT, element}))
-        LOG_ERROR(log, "SystemLog queue is full");
+    if (queue.size() >= DBMS_SYSTEM_LOG_QUEUE_SIZE / 2)
+    {
+        // The queue more than half full, time to flush.
+        const uint64_t queue_end = queue_front_index + queue.size();
+        if (requested_flush_before < queue_end)
+        {
+            requested_flush_before = queue_end;
+        }
+        flush_event.notify_all();
```

Comment:
Only the background thread + synchronous flush requests wait on flush_event, and there is not so many of them, so waking them all should be cheap. I'd like to avoid adding extra moving parts unless absolutely necessary.

Switching to a single notification makes sense, will do.